execution/registry/registry.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2017 Monax Industries Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registry
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hyperledger/burrow/crypto"
+	"github.com/tendermint/tendermint/p2p"
+)
+
+// NewRegisteredNode takes a p2p node key, and returns a node entry with a signature
+func NewRegisteredNode(nodeKey *p2p.NodeKey, moniker, netAddress string) (*RegisteredNode, []byte, error) {
+	burrowNodePubKey, err := crypto.PublicKeyFromTendermintPubKey(nodeKey.PubKey())
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not convert tm publicKey: %v", err)
+	}
+
+	burrowNodeID, err := crypto.AddressFromHexString(string(nodeKey.ID()))
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not convert tm nodeID: %v", err)
+	}
+
+	node := &RegisteredNode{
+		Moniker:    moniker,
+		NetAddress: netAddress,
+		ID:         burrowNodeID,
+		PublicKey:  burrowNodePubKey,
+	}
+
+	data, err := node.Marshal()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sig, err := nodeKey.PrivKey.Sign(data)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return node, sig, nil
+}
+
+func AuthorizedPeersProvider(state IterableReader) func() (authorizedPeersID []string, authorizedPeersAddress []string) {
+	return func() ([]string, []string) {
+		var authorizedPeersID, authorizedPeersAddress []string
+
+		state.IterateNodes(func(addr crypto.Address, node *RegisteredNode) error {
```

Comment:
We should check the error here.

But actually I think this is the wrong approach. In the ABCI app we get given an explicit query for a peer - we shouldn't have to enumerate all permitted nodes each time we query, we should call `GetNode` on the registry. So let's change the signature of `authorisedPeersProvider` to a query function. You may or may not find some inspiration here: https://github.com/cosmos/cosmos-sdk/blob/master/baseapp/baseapp.go#L599

This is complicated by the fact we need to be able to lookup by ID or netaddress. let's discuss this.